sotodlib/site_pipeline/make_position_match.py

```diff
@@ -0,0 +1,883 @@
+                  Should be a (2, n) or (3, n) array with columns: x, y, pol.
+
+        priors: Priors to apply when matching.
+                Should be be a n by n array where n is the number of points.
+                The priors[i, j] is the prior on the i'th point in template matching the j'th point in focal_plane.
+
+        out_thresh: Threshold at which points will be considered outliers.
+                    Should be in range [0, 1) and is checked against the
+                    probability that a point matches its mapped point in the template.
+
+        bias_lines: Include bias lines in matching.
+
+        reverse: Reverse direction of match.
+
+        vis: If true generate plots to watch the matching process.
+             Should only be used for debugging with human interaction.
+
+        cpd_args: Dictionairy containing kwargs to be passed into AffineRegistration.
+                  See the pycpd docs for what these can be.
+    Returns:
+
+        mapping: Mapping between elements in template and focal_plane.
+                 focal_plane[i] = template[mapping[i]]
+
+        outliers: Indices of points that are outliers.
+                  Note that this is in the basis of mapping and focal_plane, not template.
+
+        P: The likelihood array without priors applied.
+
+        TY: The transformed points.
+    """
+    if not bias_lines:
+        focal_plane = focal_plane[:, 1:]
+        template = template[:, 1:]
+    if reverse:
+        cpd_args.update({"X": focal_plane, "Y": template})
+    else:
+        cpd_args.update({"Y": focal_plane, "X": template})
+    reg = AffineRegistration(**cpd_args)
+
+    if vis:
+        frames = []
+
+        def store_frames(frames, iteration, error, X, Y):
+            frames += [[iteration, error, X, Y]]
+
+        fig = plt.figure()
+        fig.add_axes([0, 0, 1, 1])
+        callback = partial(store_frames, frames=frames)
+        reg.register(callback)
+        anim = ani.FuncAnimation(
+            fig=fig,
+            func=partial(
+                visualize, frames=frames, ax=fig.axes[0], bias_lines=bias_lines
+            ),
+            frames=len(frames),
+            interval=200,
+        )
+        plt.show()
+    else:
+        reg.register()
+
+    P = reg.P.T
+    if reverse:
+        P = reg.P
+    TY = reg.TY
+    if not bias_lines:
+        TY = np.column_stack((np.zeros(len(TY)), TY))
+
+    if priors is None:
+        priors = 1
+
+    # Solve the assignment problem
+    row_ind, col_ind = linear_sum_assignment(P * priors, True)
+    if len(row_ind) < len(focal_plane):
+        mapping = np.argmax(P * priors, axis=0)
+        mapping[col_ind] = row_ind
+    else:
+        mapping = row_ind[np.argsort(col_ind)]
+
+    outliers = np.array([])
+    if out_thresh > 0:
+        outliers = np.where(P[mapping, range(P.shape[1])] < out_thresh)[0]
+
+    return mapping, outliers, P, TY
+
+
+def main():
```

Comment:
This is a monster of a function.  It needs more narrative comments, breaking things up and explaining what subblocks are doing.

Please take some of the little stand-alone blocks (such as the bit that assigns each det to a bias_group) and make then separate functions.  It may seem silly to have a function that is only called once, but the side benefits are:
- any local variables do not stick around forever in the main scope -- it's easier to tell what things are important to carry forward
- it encourages use of cohesive data structures that are easy to pass in and out of functions (e.g. dicts or AxisMans instead of a bunch of named arrays of the same size)
- it is easier to appreciate the flow of the main function, as a whole.  (A well chosen function name will document what the function is doing, so you don't even have to comment about it.)

In such small, stand-alone functions ... don't worry about docstrings; mark them private by naming them with a leading `_`.